File: frontend/src/services/composio.ts
/**
 * Composio API Client
 * 
 * Provides type-safe API functions for all Composio operations.
 * Uses axios for HTTP requests and follows TanStack Query patterns.
 */

import axios, { AxiosError } from 'axios';

const API_BASE = '/api/v1/composio';

// ============================================================================
// Types & Interfaces
// ============================================================================

export interface AuthScheme {
  type: 'oauth2' | 'oauth1' | 'api_key' | 'basic_auth' | 'bearer_token';
  fields?: {
    required: string[];
    optional: string[];
  };
}

export interface Toolkit {
  slug: string;
  name: string;
  description: string;
  logo_url?: string;
  categories: string[];
  auth_schemes: AuthScheme[];
}

export interface ToolParameter {
  type: string;
  description?: string;
  required?: boolean;
  default?: any;
  examples?: any[];
  properties?: Record<string, ToolParameter>;
  items?: ToolParameter;
}

export interface ToolSchema {
  type: string;
  properties: Record<string, ToolParameter>;
  required: string[];
}

export interface ToolDefinition {
  name: string;
  slug: string;
  description: string;
  input_parameters: ToolSchema;
  output_parameters: ToolSchema;
  scopes: string[];
  no_auth: boolean;
  version?: string;
}

export interface Connection {
  connection_id: string;
  app_type: string;
  status: 'ACTIVE' | 'PENDING' | 'FAILED' | 'EXPIRED';
  created_at?: string;
  updated_at?: string;
}

export interface ConnectionInitiateResponse {
  connection_id: string;
  auth_url: string;
  toolkit_slug: string;
  status: string;
}

export interface ConnectionStatusResponse {
  connection_id: string;
  status: 'ACTIVE' | 'PENDING' | 'FAILED' | 'EXPIRED';
  message?: string;
}

export interface ToolkitListResponse {
  success: boolean;
  data: Toolkit[];
  message: string;
}

export interface ToolDefinitionListResponse {
  success: boolean;
  data: ToolDefinition[];
  message: string;
}

export interface ConnectionListResponse {
  success: boolean;
  data: Connection[];
  message: string;
}

export interface ToolExecutionRequest {
  tool_slug: string;
  user_id: string;
  parameters: Record<string, any>;
}

export interface ToolExecutionResponse {
  success: boolean;
  result?: any;
  error?: string;
  tool_slug: string;
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Fetch all available toolkits with optional filtering
 */
export const fetchToolkits = async (params?: {
  category?: string | null;
  search?: string | null;
}): Promise<ToolkitListResponse> => {
  const queryParams = new URLSearchParams();
  if (params?.category) queryParams.append('category', params.category);
  if (params?.search) queryParams.append('search', params.search);

  const response = await axios.get<ToolkitListResponse>(
    `${API_BASE}/toolkits${queryParams.toString() ? `?${queryParams}` : ''}`
  );
  return response.data;
};

/**
 * Fetch tool definitions for a specific toolkit
 */
export const fetchToolkitTools = async (
  toolkitSlug: string
): Promise<ToolDefinitionListResponse> => {
  const response = await axios.get<ToolDefinitionListResponse>(
    `${API_BASE}/toolkits/${toolkitSlug}/tools`
  );
  return response.data;
};

/**
 * Fetch metadata for a specific toolkit
 */
export const fetchToolkitMetadata = async (toolkitSlug: string): Promise<any> => {
  const response = await axios.get(`${API_BASE}/toolkits/${toolkitSlug}/metadata`);
  return response.data;
};

/**
 * Initiate OAuth connection for a toolkit
 */
export const initiateConnection = async (
  toolkitSlug: string,
  userId: string
): Promise<ConnectionInitiateResponse> => {
  const response = await axios.post(`${API_BASE}/connections/initiate`, {
    toolkit_slug: toolkitSlug,
    user_id: userId,
  });
  return response.data.data;
};

/**
 * Check the status of a connection
 */
export const checkConnectionStatus = async (
  connectionId: string
): Promise<ConnectionStatusResponse> => {
  const response = await axios.get(
    `${API_BASE}/connections/${connectionId}/status`
  );
  return response.data.data;
};

/**
 * Fetch all connections for a specific user
 */
export const fetchUserConnections = async (
  userId: string
): Promise<ConnectionListResponse> => {
  const response = await axios.get<ConnectionListResponse>(
    `${API_BASE}/users/${userId}/connections`
  );
  return response.data;
};

/**
 * Disconnect an app connection
 */
export const disconnectApp = async (connectionId: string): Promise<void> => {
  await axios.delete(`${API_BASE}/connections/${connectionId}`);
};

/**
 * Execute a Composio tool
 */
export const executeTool = async (
  request: ToolExecutionRequest
): Promise<ToolExecutionResponse> => {
  const response = await axios.post(`${API_BASE}/tools/execute`, request);
  return response.data.data;
};

/**
 * Get available categories for filtering toolkits
 */
export const fetchCategories = async (): Promise<string[]> => {
  // This would need to be implemented on the backend
  // For now, return common categories
  return [
    'Communication',
    'Productivity',
    'Development',
    'Marketing',
    'Sales',
    'Support',
    'Finance',
    'HR',
    'Analytics',
  ];
};

// ============================================================================
// Error Handling
// ============================================================================

export interface ApiError {
  error: string;
  message: string;
  [key: string]: any;
}

export const isApiError = (error: unknown): error is AxiosError<ApiError> => {
  return axios.isAxiosError(error) && error.response !== undefined;
};

export const getErrorMessage = (error: unknown): string => {
  if (isApiError(error)) {
    return error.response?.data?.message || error.message;
  }
  if (error instanceof Error) {
    return error.message;
  }
  return 'An unknown error occurred';
};

File: agent_dojo/schemas/composio.py
"""Pydantic schemas for Composio integration"""

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class AuthSchemeType(str, Enum):
    """Authentication scheme types"""

    OAUTH2 = "oauth2"
    OAUTH1 = "oauth1"
    API_KEY = "api_key"
    BASIC_AUTH = "basic_auth"
    BEARER_TOKEN = "bearer_token"


class AuthSchemeFields(BaseModel):
    """Authentication scheme field requirements"""

    required: list[str] = Field(default_factory=list)
    optional: list[str] = Field(default_factory=list)


class AuthScheme(BaseModel):
    """Authentication scheme information"""

    type: AuthSchemeType
    fields: AuthSchemeFields | None = None


class ToolkitInfo(BaseModel):
    """Toolkit (App) information"""

    slug: str = Field(..., description="Unique toolkit identifier")
    name: str = Field(..., description="Display name")
    description: str = Field(..., description="Toolkit description")
    logo_url: str | None = Field(None, description="Logo URL")
    categories: list[str] = Field(default_factory=list)
    auth_schemes: list[AuthScheme] = Field(default_factory=list)


class ToolParameter(BaseModel):
    """Tool parameter definition"""

    type: str = Field(..., description="Parameter type (string, number, object, etc.)")
    description: str | None = Field(None, description="Parameter description")
    required: bool = Field(False, description="Whether parameter is required")
    default: Any | None = Field(None, description="Default value")
    examples: list[Any] = Field(default_factory=list)
    properties: dict[str, "ToolParameter"] | None = Field(
        None, description="Nested properties for object types"
    )
    items: "ToolParameter | None" = Field(None, description="Array item schema")


class ToolSchema(BaseModel):
    """Tool input/output schema"""

    type: str = Field(default="object")
    properties: dict[str, ToolParameter] = Field(default_factory=dict)
    required: list[str] = Field(default_factory=list)


class ToolDefinition(BaseModel):
    """Raw tool definition with complete metadata"""

    name: str = Field(..., description="Tool display name")
    slug: str = Field(..., description="Tool slug/identifier")
    description: str = Field(..., description="Tool description")
    input_parameters: ToolSchema = Field(
        default_factory=ToolSchema, description="Input parameter schema"
    )
    output_parameters: ToolSchema = Field(
        default_factory=ToolSchema, description="Output parameter schema"
    )
    scopes: list[str] = Field(
        default_factory=list, description="Required OAuth scopes"
    )
    no_auth: bool = Field(False, description="Whether authentication is required")
    version: str | None = Field(None, description="Tool version")


class ConnectionStatus(str, Enum):
    """Connection status enum"""

    ACTIVE = "ACTIVE"
    PENDING = "PENDING"
    FAILED = "FAILED"
    EXPIRED = "EXPIRED"


class ConnectionRequest(BaseModel):
    """Request to initiate a connection"""

    toolkit_slug: str = Field(..., description="Toolkit to connect")
    user_id: str = Field(..., description="User ID for the connection")


class ConnectionInitiateResponse(BaseModel):
    """Response from connection initiation"""

    connection_id: str = Field(..., description="Connection ID")
    auth_url: str = Field(..., description="OAuth authorization URL")
    toolkit_slug: str = Field(..., description="Toolkit slug")
    status: ConnectionStatus = Field(..., description="Connection status")


class ConnectionInfo(BaseModel):
    """User connection information"""

    connection_id: str = Field(..., description="Unique connection ID")
    app_type: str = Field(..., description="App/toolkit type")
    status: ConnectionStatus = Field(..., description="Connection status")
    created_at: datetime | None = Field(None, description="Connection creation time")
    updated_at: datetime | None = Field(None, description="Last update time")


class ConnectionStatusResponse(BaseModel):
    """Response for connection status check"""

    connection_id: str = Field(..., description="Connection ID")
    status: ConnectionStatus = Field(..., description="Current status")
    message: str | None = Field(None, description="Status message")


class ToolExecutionRequest(BaseModel):
    """Request to execute a tool"""

    tool_slug: str = Field(..., description="Tool slug to execute")
    user_id: str = Field(..., description="User ID")
    parameters: dict[str, Any] = Field(..., description="Tool parameters")


class ToolExecutionResponse(BaseModel):
    """Response from tool execution"""

    success: bool = Field(..., description="Whether execution was successful")
    result: Any | None = Field(None, description="Execution result")
    error: str | None = Field(None, description="Error message if failed")
    tool_slug: str = Field(..., description="Executed tool slug")


class ToolkitListResponse(BaseModel):
    """Response containing list of toolkits"""

    success: bool = True
    data: list[ToolkitInfo]
    message: str = "Toolkits retrieved successfully"


class ToolDefinitionListResponse(BaseModel):
    """Response containing list of tool definitions"""

    success: bool = True
    data: list[ToolDefinition]
    message: str = "Tool definitions retrieved successfully"


class ConnectionListResponse(BaseModel):
    """Response containing list of user connections"""

    success: bool = True
    data: list[ConnectionInfo]
    message: str = "Connections retrieved successfully"


# Update forward references for recursive models
ToolParameter.model_rebuild()
